### ISSUE TITLE
test_phyml_tool.py failure Phylip/interlaced2.phy_phyml_tree.txt

### ISSUE THREAD

**Issue by blaiseli** [biopython/biopython#919]
I saw on the biopython mailing list that tests were welcome before the release of biopython 1.68 on pypi.

For me, the build seemed to work, but the tests ended as follows:

```
======================================================================
ERROR: test_nuccore_X52960 (test_SeqIO_online.EntrezTests)
Bio.Entrez.efetch('nuccore', id='X52960', ...)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_SeqIO_online.py", line 79, in <lambda>
    method = lambda x: x.simple(d, f, e, l, c)
  File "test_SeqIO_online.py", line 56, in simple
    handle = Entrez.efetch(db=database, id=entry, rettype=f, retmode="text")
  File "/home/bli/src/biopython-1.68/build/lib.linux-x86_64-2.7/Bio/Entrez/__init__.py", line 180, in efetch
    return _open(cgi, variables, post=post)
  File "/home/bli/src/biopython-1.68/build/lib.linux-x86_64-2.7/Bio/Entrez/__init__.py", line 526, in _open
    raise exception
HTTPError: HTTP Error 502: Bad Gateway

======================================================================
ERROR: test_phyml (test_phyml_tool.AppTests)
Run PhyML using the wrapper.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_phyml_tool.py", line 55, in test_phyml
    tree = Phylo.read(EX_PHYLIP + '_phyml_tree.txt', 'newick')
  File "/home/bli/src/biopython-1.68/build/lib.linux-x86_64-2.7/Bio/Phylo/_io.py", line 63, in read
    tree = next(tree_gen)
  File "/home/bli/src/biopython-1.68/build/lib.linux-x86_64-2.7/Bio/Phylo/_io.py", line 50, in parse
    with File.as_handle(file, 'r') as fp:
  File "/home/bli/lib/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "/home/bli/src/biopython-1.68/build/lib.linux-x86_64-2.7/Bio/File.py", line 88, in as_handle
    with open(handleish, mode, **kwargs) as fp:
IOError: [Errno 2] No such file or directory: 'Phylip/interlaced2.phy_phyml_tree.txt'

----------------------------------------------------------------------
Ran 264 tests in 458.958 seconds

FAILED (failures = 2)
```

I use a locally compiled python 2.7.12, on Xubuntu 16.04

**Comment by peterjc** [comment]
Thanks for getting in touch.

The ``HTTP Error 502: Bad Gateway`` is a temporary outage from the NCBI and/or your network, most likely it will work on rerunning.

The PHYML failure from ``test_phyml_tool.py`` suggests something annoyingly different in the version of the command line tool you have installed - can you check what version you have please?

(Neither worries me in terms of holding up the release)

**Comment by blaiseli** [comment]
I had a manually installed phyml which indicated a lower version date as the globally installed phyml packaged by Ubuntu. I removed the locally-installed version and re-ran the tests now that I'm sure that phyml is in the Ubuntu-packaged one, which says it is "version 20160207" (I'm not sure this date is meaningful, though).

The error happens again (together with another internet-related error):
```
======================================================================
ERROR: test_search (test_SCOP_online.ScopSearch)
Bio.SCOP.search(...)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_SCOP_online.py", line 21, in test_search
    handle = SCOP.search("1JOY")
  File "/home/bli/src/biopython-1.68/build/lib.linux-x86_64-2.7/Bio/SCOP/__init__.py", line 871, in search
    return _open(cgi, variables)
  File "/home/bli/src/biopython-1.68/build/lib.linux-x86_64-2.7/Bio/SCOP/__init__.py", line 892, in _open
    handle = urlopen(cgi)
  File "/home/bli/lib/python2.7/urllib2.py", line 154, in urlopen
    return opener.open(url, data, timeout)
  File "/home/bli/lib/python2.7/urllib2.py", line 429, in open
    response = self._open(req, data)
  File "/home/bli/lib/python2.7/urllib2.py", line 447, in _open
    '_open', req)
  File "/home/bli/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/home/bli/lib/python2.7/urllib2.py", line 1228, in http_open
    return self.do_open(httplib.HTTPConnection, req)
  File "/home/bli/lib/python2.7/urllib2.py", line 1198, in do_open
    raise URLError(err)
URLError: <urlopen error [Errno 110] Connection timed out>

======================================================================
ERROR: test_phyml (test_phyml_tool.AppTests)
Run PhyML using the wrapper.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_phyml_tool.py", line 55, in test_phyml
    tree = Phylo.read(EX_PHYLIP + '_phyml_tree.txt', 'newick')
  File "/home/bli/src/biopython-1.68/build/lib.linux-x86_64-2.7/Bio/Phylo/_io.py", line 63, in read
    tree = next(tree_gen)
  File "/home/bli/src/biopython-1.68/build/lib.linux-x86_64-2.7/Bio/Phylo/_io.py", line 50, in parse
    with File.as_handle(file, 'r') as fp:
  File "/home/bli/lib/python2.7/contextlib.py", line 17, in __enter__
    return self.gen.next()
  File "/home/bli/src/biopython-1.68/build/lib.linux-x86_64-2.7/Bio/File.py", line 88, in as_handle
    with open(handleish, mode, **kwargs) as fp:
IOError: [Errno 2] No such file or directory: 'Phylip/interlaced2.phy_phyml_tree.txt'

----------------------------------------------------------------------
Ran 264 tests in 706.997 seconds

FAILED (failures = 2)
```

Just to be sure, I erased the biopython-1.68 directory and re-started from the unpacking and building before running the tests again.
The test for phyml failed again (along with 2 internet-related failures: one connection timed out for `test_SCOP_online.py` and one 404 error happend for `test_TogoWS.py`).

**Comment by blaiseli** [comment]
There is a `Tests/Phylip/interlaced2.phy_phyml_tree` file, with no ".txt".
Some testing confirms that my version of phyml names the output tree without ".txt"

**Comment by peterjc** [comment]
It looks like neither our TravisCI setup nor the main Linux buildbot have phyml installed, although the test is massing on our Mac OS X buildbot, e.g.

http://testing.open-bio.org/biopython/builders/OS%20X%20-%20Python%202.7/builds/641/steps/shell/logs/stdio

I'll check what version that has installed, but it sounds like the test needs to look for the output file with and without the ``.txt``

CC @etal

**Comment by peterjc** [comment]
I just installed PhyML 20120412 for 64-bit Mac OS X on my laptop and the Biopython test worked - this seems to be the latest version from http://www.atgc-montpellier.fr/phyml/download.php 

```
$ curl -L -O http://www.atgc-montpellier.fr/download/binaries/phyml/PhyML-3.1.zip
$ unzip PhyML-3.1.zip
$ ./PhyML-3.1/PhyML-3.1_macOS-MountainLion --version


. This is PhyML version 20120412.


$ mv PhyML-3.1/PhyML-3.1_macOS-MountainLion ~/bin/phyml
```

From the Biopython ``Tests/`` directory,

```
$ python run_tests.py test_phyml_tool.py
Python version: 2.7.10 (default, Oct 23 2015, 19:19:21) 
[GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.0.59.5)]
Operating system: posix darwin
test_phyml_tool ... ok
----------------------------------------------------------------------
Ran 1 test in 0.737 seconds


```

**Comment by blaiseli** [comment]
The output of `phyml --version` is really misleading, and Ubuntu is actually packaging an older version.

A version 4.0 seems to be in preparation, as I can infer from this job offer: <http://www.sfbi.fr/content/ingenieur-bioinformatique-evolutive-0>

I hope the authors will make the `--version` option better behaving.

**Comment by peterjc** [comment]
It seems the Ubuntu package is very old, currently only PhyML 2.2

http://packages.ubuntu.com/precise/science/phyml

However Debian have packaged PhyML 3.2

https://packages.debian.org/source/sid/phyml

Ubuntu should really catch up with the Debian release here (and then we can ask TravisCI to whitelist the package for us to use in testing).

Meanwhile I only found PhyML 3.1 on the group's website, so I wonder where Debian got PhyML 3.2 from...

**Comment by blaiseli** [comment]
Newer versions of Ubuntu seem to package newer versions of phyml:
<http://packages.ubuntu.com/yakkety/science/phyml>
The version I have should be this thing they call 3.2.

I found the development version here: <https://github.com/stephaneguindon/phyml>.

I installed it:

```
git clone https://github.com/stephaneguindon/phyml.git
cd phyml/
./autogen.sh
./configure --prefix=${HOME}
make
make install
# To get the new local version taken into account:
hash phyml
```

Trying the `--version` option:

```
$ phyml --version


. Command line: phyml --version 

. This is PhyML version 3.3.20160826.

```

Trying to run it got me a "Segmentation fault", but generated two empty files with the following suffix: "_phyml_tree" and "_phyml_stats", without ".txt" (like with the version packaged for Ubuntu 16.04).

My bet is that the version packaged in the recent Debian and Ubuntu releases is more recent than the 3.1 available for download on the ATGC site.

I don't know what is the more likely case for the typical Biopython user: manual download from the ATGC site or install from a distribution's package ?

**Comment by peterjc** [comment]
Well spotted, https://github.com/stephaneguindon/phyml has tags for v3.2 and two date-stamped v3.2.yyymmdd releases. That probably is where Debian/Ubuntu are getting things from.

I was looking at Ubuntu Precise as that is what TravisCI uses currently.

As to where Biopython users get PhyML from, probably a mixture. What do you think @etal?

In any case, updating our test to try with and without the ``.txt`` seems best.

**Comment by peterjc** [comment]
I've opened a request for TravisCI to add phyml to their whitelist to simplify testing it there:
https://github.com/travis-ci/apt-package-whitelist/issues/3317

**Comment by peterjc** [comment]
I built the version from git under CentOS with a tiny tweak to ``configure``,

```
-am__api_version='1.14'
+am__api_version='1.11'
```

The good news is (as you reported above) it has a useful self-reported version number. It also seg-faulted for me too.

I tried building the tagged v3.2.0, v3.2.20160529, and v3.2.20160530 but got ``configure.ac:69: required file `./ltmain.sh' not found``.

**Comment by peterjc** [comment]
I've included PhyML 3.1 on the TravisCI system so it runs ``test_phyml_tool.py`` https://github.com/biopython/biopython/commit/006f869f10cb6a2f6ab75909e4173f950ec5c271

**Comment by peterjc** [comment]
OK, the seg-fault (issue #921) is fixed as of PhyML 3.3.20160826 by https://github.com/stephaneguindon/phyml/commit/03a2597eeeaf4877cb25cf76f2fc814c2ba8e2c7?w=1

Using this we still have the extension problem to deal with:

```
$ python3.5 test_phyml_tool.py
test_phyml (__main__.AppTests)
Run PhyML using the wrapper. ... ERROR

======================================================================
ERROR: test_phyml (__main__.AppTests)
Run PhyML using the wrapper.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_phyml_tool.py", line 55, in test_phyml
    tree = Phylo.read(EX_PHYLIP + '_phyml_tree.txt', 'newick')
  File "/home_local/buildslave/lib/python3.5/site-packages/Bio/Phylo/_io.py", line 63, in read
    tree = next(tree_gen)
  File "/home_local/buildslave/lib/python3.5/site-packages/Bio/Phylo/_io.py", line 50, in parse
    with File.as_handle(file, 'r') as fp:
  File "/home_local/buildslave/lib/python3.5/contextlib.py", line 59, in __enter__
    return next(self.gen)
  File "/home_local/buildslave/lib/python3.5/site-packages/Bio/File.py", line 88, in as_handle
    with open(handleish, mode, **kwargs) as fp:
FileNotFoundError: [Errno 2] No such file or directory: 'Phylip/interlaced2.phy_phyml_tree.txt'

----------------------------------------------------------------------
Ran 1 test in 0.694s

FAILED (errors=1)
```

The alternatively named output file does exist:

```
$ more Phylip/interlaced2.phy_phyml_tree
(IXI_236:0.08876281,IXI_237:0.04857371,(IXI_235:0.00000001,IXI_234:0.00000001)0.000000:0.00000001);
```
